### PR TITLE
[openwebnet] Add a time stamp when an alarm zone event occurs

### LINKED ISSUE #14535: [openwebnet] Alarm: add a time stamp when a alarm zone event occurs
Channel CHANNEL_ALARM_ZONE_ALARM_STATE for an active zone should show  and persist the time of the event  

**Expected Behavior**

MM/DD/YY,HH:MM INTRUSION

### PR DESCRIPTION
**Pull request by fabgio**
**Description:**
This PR allows to show a time stamp when an alarm zone event occurs on Bticino/Legrand Alarm systems such as an Intrusion or a Tampering This PR closes #14535

**Test**
Tested on BTicino/Legrand bulgrar-alarm unit 3486

### CODE REVIEW

File: bundles/org.openhab.binding.openwebnet/README.md
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/README.md:230]: can you modifiy only the lines related to adding the new channel?

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/OpenWebNetBindingConstants.java
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/OpenWebNetBindingConstants.java:176]: suggest to call it CHANNEL_ALARM_ZONE_ALARM_TIMESTAMP

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:243]: you have to choose when updateZoneAlarmTimeStamp is called. This way it's called 2 times: in the switch inside updateZoneAlarm and then again in the updateZone switch

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:247]: see previous comment

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:251]: see previous comment

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:255]: see previous comment

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:263]: see previous comment

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:259]: see previous comment

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:228]: you have to choose when updateZoneAlarmTimeStamp is called. This way it's called 2 times: in the switch inside updateZoneAlarm and then again in the updateZone switch

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:273]: this line is useless. Why don't you do the same like in initialize():
`updateState(CHANNEL_ALARM_ZONE_TIMESTAMP, ....`

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusAlarmZone.xml
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusAlarmZone.xml:20]:
```suggestion
			<channel id="timestamp" typeId="zoneAlarmTimestamp"/>
```
Timestamp and not TimeStamp (it's one word)

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:464]:
```suggestion
	<channel-type id="zoneAlarmTimestamp">
```

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:466]:
```suggestion
		<label>Zone Alarm Timestamp</label>
```

**Review by mvalla — CHANGES_REQUESTED** [review]
@fabgio see required changes.

You also have to generate the default English localization, see: https://github.com/openhab/openhab-addons#translations

This is a second PR after #14743
PS
Do not change again the PR, beacuse each time a new review is required.

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/OpenWebNetBindingConstants.java
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/OpenWebNetBindingConstants.java:176]: done

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:243]: You are right, I  choose to call it in updateZone() to invoke it only once

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:273]: at this point could I call updateTimeStmp()?

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:243]: done

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:247]: done

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:251]: done

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:263]: done

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:259]: done

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:228]: ok I will call it in updateZone() only

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusAlarmZone.xml
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/BusAlarmZone.xml:20]: done

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:464]: done

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:466]: done

File: bundles/org.openhab.binding.openwebnet/README.md
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/README.md:230]: Oh my mistake I had a prooblem in git conflict's resolution I will push the right file

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/i18n/openwebnet.properties
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/i18n/openwebnet.properties:288]:
```suggestion
channel-type.openwebnet.zoneAlarmTimestamp.label = Zone Alarm Timestamp
```

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:466]:
```suggestion
		<label>Zone Alarm Timestamp</label>
```

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:266]:
```suggestion
```

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (mvalla) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:267]:
```suggestion
        updateState(CHANNEL_ALARM_ZONE_ALARM_TIMESTAMP, new DateTimeType());
```

File: bundles/org.openhab.binding.openwebnet/README.md
Comment (mvalla) [reply bundles/org.openhab.binding.openwebnet/README.md:230]: maybe spaces? Copy paste the original lines from original version

**Review by mvalla — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/i18n/openwebnet.properties
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/i18n/openwebnet.properties:288]: done

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:267]: line 266 is unuseful at this point so  I would remove it and  as a consequence   would  updateZoneTimeStamp() implementation still convenient? I could  simply invoke  updateState(CHANNEL_ALARM_ZONE_ALARM_TIMESTAMP, new DateTimeType())  directly at line 228...

Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java]: done as per previous comment

**Review by mvalla — APPROVED** [review]
LGTM

**Comment by fabgio** [comment]
Thank you so much!

**Comment by lolodomo** [comment]
Please add migration instructions for the new channel. 
A link to the documentation
https://next.openhab.org/docs/developer/bindings/thing-xml.html#updating-thing-types

**Review by mvalla — CHANGES_REQUESTED** [review]
I never used such migration feature, but I think that:
- directory is wrong
- <label> should be the new channel label, not its description
- according to instructions this is also needed that _"In addition to the update instructions, the thing-type definition needs to add a property thingTypeVersion to prevent newly created things from being modified"_

**Comment by fabgio** [comment]
@mvalla I see, I will check accordingly to your tips. 

> I never used such migration feature

So do I, that is completely new to me

**Comment by fabgio** [comment]
@mvalla I fulfilled all of the three points you mentioned above. Moreover I validated update.xml for correctness

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml
Comment (lolodomo) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:464]: You could adds `<category>Time</category>`

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (lolodomo) [inline bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:134]: Please remove this blank line you added

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/update/update.xml
Comment (lolodomo) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/update/update.xml:10]: No need to set a label, it will be taken from the channel type.
Requires openhab/openhab-core#3576

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH.INF/update/update.xml
Comment (lolodomo) [inline bundles/org.openhab.binding.openwebnet/src/main/resources/OH.INF/update/update.xml:1]: File and folders folders OH.INFO/update to be removed.

**Review by lolodomo — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/java/org/openhab/binding/openwebnet/internal/handler/OpenWebNetAlarmHandler.java:134]: Done

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/update/update.xml
Comment (fabgio) [reply bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/update/update.xml:10]: OK. Label tag removed

**Comment by fabgio** [comment]
Do you mean  I should remove that line,right? I removed it and validate update.xml with no errors

File: bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml
Comment (lolodomo) [reply bundles/org.openhab.binding.openwebnet/src/main/resources/OH-INF/thing/channels.xml:464]: You have to add it at the right place inside <channel-type>. Look at an example in existing code.

**Comment by lolodomo** [comment]
You have update.xml twice in 2 different folders: OH-INF/update and OH.INF/update. Only the first is needed.

**Comment by fabgio** [comment]
I am striving to get rid of the wrong OH.INF dir from my origin even by forcing push... I will handle this problem

**Comment by fabgio** [comment]
ok OH.INF dir has gone now!!!

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Comment by lolodomo** [comment]
DCO KO.
Maybe you did not sign-off properly one of your last commits ?
Please fix it in case it is the last one. If not, squash several of your commits.

**Comment by fabgio** [comment]
DTO is ok now.. tommorow I will do further integration-tests, waiting for @mvalla review

**Comment by fabgio** [comment]
I appreciated your support! Thanks a lot

**Comment by lolodomo** [comment]
> tommorow I will do further integration-tests, waiting for @mvalla review

Do you mean the PR is not yet ready for a merge?

**Comment by lolodomo** [comment]
Current code looks good to me.

**Comment by fabgio** [comment]
@lolodomo, i did tests. To me is  ready for merge

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Comment by fabgio** [comment]
Thanks again See you soon!

**Comment by fabgio** [comment]
Hi lolodomo! After the merging of my PR I am not able to compile the binding anymore. Even Cloning from Main repo! This is the message

 

[WARNING] Could not transfer metadata org.openhab:openhab-super-pom/maven-metadata.xml from/to github (https://maven.pkg.github.com/fabgio/14535_openwebnet_4.0.0_publish): authentication failed for https://maven.pkg.github.com/fabgio/14535_openwebnet_4.0.0_publish/org/openhab/openhab-super-pom/maven-metadata.xml, status: 401 Unauthorized

 

It looks that maven looks for  an unwanted dependency…  I don’t know where it comes from I need help.. I am terrified!!!  

 

Please help!!

**Pull request merged**
